Implement special treatment for own chat

marhkb opened the pull request:
![own_chat](https://user-images.githubusercontent.com/3630213/142251105-4b2c2063-5c12-40e4-a007-78fc3253f441.png)

fixes #40

melix99 commented on src/session/sidebar/row.rs:
```diff
@@ -167,8 +171,91 @@ impl Row {
+                                if args[1].get::<bool>().unwrap() {
+                                    adw::Avatar::builder()
+                                        .size(44)
+                                        .icon_name("document-save-symbolic")
```
```suggestion
                                        .icon_name("user-bookmarks-symbolic")
```
![Screenshot from 2021-11-17 19-15-11](https://user-images.githubusercontent.com/23530586/142258595-edacae9a-9e20-47a7-8615-fece58d1a8d2.png)

marhkb committed:
sidebar: Enable searching for own chat

This commit allows to search for the own chat in the sidebar. To find
the own chat, the query string must be contained in the string "Saved
Messages" (translated into the respective local language).

This commit is the starting point for treating the own chat differently.
Further commits will follow.

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -163,6 +169,51 @@ impl ChatHistory {
+
+            match interlocutor_id(chat) {
+                Some(interlocutor_id) => {
+                    let session_expression = gtk::PropertyExpression::new(
```
Creating expressions like this creates the problem that we had with sidebar-row: expressions won't unwatch themself automatically when new ones are created after another call to `set_chat()`.

We have three ways to fix this:
1. Do what I did in sidebar-row (basically storing all the bindings and unwatching before creating new ones);
2. Make the expressions autoupdate themselves (not always possible). To do that we have to create the expressions just a single time (ideally at `constructed()`) and binding to the internal property for making them update themselves (in our case the `chat` property);
3. Use manual code instead of expressions. That ofc means that we won't get updates like expressions do.

For this case I suggest using 1. 2 could also work, but I think it can become too complicated (we'd be forced to have a bind of `me_expression` even with non-private chats).

marhkb replied:
As suggested, I applied the first solution.

melix99 commented on src/session/sidebar/mod.rs:
```diff
@@ -197,15 +200,35 @@ impl Sidebar {
+                                chats.chat_ids
+                                    .into_iter()
+                                    // The own user id is the same as the own chat id, so we can
+                                    // chan this here.
```
```suggestion
                                    // chain this here.
```

melix99 commented on data/resources/ui/content-message-row.ui:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Mh? I guess you added this file by accident?

marhkb replied:
Oops, that somehow crept in. I guess while rebasing? I really don't know.

melix99 replied:
I was getting this problem all the time and I couldn't figure out what it was, then recently I realized that it was Gnome Builder's auto saving option (enabled by default). So, if you use Builder that could be the reason. Otherwise it was there from the rebase.

melix99 commented on src/session/sidebar/row.rs:
```diff
@@ -180,9 +183,93 @@ impl Row {
-                bindings.push(title_binding);
+                let is_own_chat_expression = match chat.type_().user().map(User::id) {
+                    Some(interlocutor_id) => {
+                        let me_expression = gtk::PropertyExpression::new(
```
I guess we really should handle the 'me' User a bit better, to avoid having complicating things unnecessarily. Is it ok for you fetching the user in the Sessions's constructor? Otherwise we can do it in another PR.

marhkb replied:
Yeah, I can try that. That was actually the approach I attempted first when implementing `Session::me()` but IIRC I got weird effects back then like users not loading, etc. I guess I did something wrong.

It's fine for me to do it in this PR.

marhkb replied:
Ok, I opened a separate PR (#208). I think this is wiser.

marhkb replied:
Ok, #208 is closed, so we handle `me` still as optional because we cannot know when it arrives.

marhkb committed:
chat: Use bookmarks icon avatar for own chat

Wherever the own user chat's avatar will be displayed, it will now al-
ways show a bluish bookmarks icon instead of initials or a custom image.

marhkb committed:
chat: Use title "Saved Messages" for own chat

melix99 commented on src/session/sidebar/row.rs:
```diff
@@ -347,11 +350,11 @@ impl Row {
         ChatType::BasicGroup(_) => true,
         ChatType::Supergroup(supergroup) => !supergroup.is_channel(),
-        ChatType::Private(_) | ChatType::Secret(_) => message.is_outgoing(),
+        ChatType::Private(_) | ChatType::Secret(_) => !is_own_chat && message.is_outgoing(),
```
We always have to show the sender in case the message is outgoing, otherwise messages forwarded in the 'Saved Messages' will not show the sender in the sidebar.

marhkb replied:
I have added the `MessageForwardInfo` for that. But there is one point where I am not satisfied. I left a `TODO` there. Maybe you have an idea.
In this turn, I cleaned up the `stringify_message` function in `row.rs`, which has become quite confusing. I hope the readability is a bit improved now.

https://github.com/melix99/telegrand/blob/a0c98c06fa1dd00d1d61035666a0bcc8d108c246/src/session/chat/message_forward_info.rs#L117-L127

marhkb replied:
I just realized the Android client does not show the sender at all. Even not for forwarded messages. However, the desktop client does.

marhkb replied:
I am not quite sure, but the crash might be caused by the fact that the message is deleted. Thus, I turned the message field into a `WeakRef`.

melix99 commented on src/session/chat/mod.rs:
```diff
@@ -494,4 +513,37 @@ impl Chat {
+        }
+    }
+
+    pub fn formatted_title_expression(&self) -> gtk::Expression {
```
What about naming it `display_name_expression`? I feel it's a clearer name.

melix99 commented on src/session/components/avatar.rs:
```diff
@@ -103,6 +109,55 @@ impl Avatar {
+            closure!(
+                |_: Self, icon_name: Option<String>, display_name: Option<String>| {
+                    icon_name
+                        .map(|_| Some(String::from("-")))
```
I think we should add a brief comment explaining why we're doing that.

marhkb committed:
sidebar-chat-row: Implement special treatment for own chat

This is that we don't show the sender name except for forwarded mes-
sages. Therefore, this commit enriches the message type with the
forward info.

In this turn, the `stringify_message` function has also been cleaned
up a bit from the spaghetti.

marhkb committed:
message-row-text: Show sender names of incoming messages for own chat

Incoming messages for the own chat can only be forwarded messages. In
this case, we show the sender name of the originator like the official
client does.

marhkb committed:
message-row: Show avatars of incoming messages for own chat

Incoming messages in the own chat can only be forwarded messages. In
this case, we show the avatar of the originator like the official client
does.
